packages/shylu/shylu_dd/frosch/src/SolverInterfaces/FROSch_MueLuPreconditioner_def.hpp

```diff
@@ -0,0 +1,129 @@
+//@HEADER
+// ************************************************************************
+//
+//               ShyLU: Hybrid preconditioner package
+//                 Copyright 2012 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Alexander Heinlein (alexander.heinlein@uni-koeln.de)
+//
+// ************************************************************************
+//@HEADER
+
+#ifndef _FROSCH_MUELUPRECONDITIONER_DEF_HPP
+#define _FROSCH_MUELUPRECONDITIONER_DEF_HPP
+
+#include <FROSch_MueLuPreconditioner_decl.hpp>
+
+
+namespace FROSch {
+
+    using namespace Teuchos;
+    using namespace Xpetra;
+
+    template<class SC,class LO,class GO,class NO>
+    int MueLuPreconditioner<SC,LO,GO,NO>::initialize()
+    {
+        FROSCH_TIMER_START_SOLVER(initializeTime,"MueLuPreconditioner::initialize");
+        this->IsInitialized_ = true;
+        this->IsComputed_ = false;
+        return 0;
+    }
+
+    template<class SC,class LO,class GO,class NO>
+    int MueLuPreconditioner<SC,LO,GO,NO>::compute()
+    {
+        FROSCH_TIMER_START_SOLVER(computeTime,"MueLuPreconditioner::compute");
+        FROSCH_ASSERT(this->IsInitialized_,"FROSch::MueLuPreconditioner: !this->IsInitialized_");
+        MueLuFactory_->SetupHierarchy(*MueLuHierarchy_);
+        MueLuHierarchy_->IsPreconditioner(false);
```

Comment:
I would say, we should be able to use MueLu in both modes. If we want to use it for the local or coarse solves, I would want to use it as a linear operator. Otherwise, we are restricted to using flexible Krylov methods for the outer iteration. Therefore, I would rather use it as a preconditioner for now. However, there are cases where it may make sense to really use it as a solver.

Since the two of us are going to change the MueLu interface in the very near anyway, I would leave it like this for now. What do you think?

Reply:
I'm fine with leaving it as is for now. Let's rename when we know more about future use cases.